### PR TITLE
Fix release forward-port post-merge review findings

### PR DESCRIPTION
**Pull request by justin808**
## Why

Follow-up to #4181 / #4162. PR #4181 is already merged, but post-merge review surfaced release-forward-port edge cases that could make the helper report the wrong plan or fail during an otherwise idempotent closeout run.

This PR keeps the original release-forward-port design and fixes the review findings without broad release-process redesign.

## What Changed

- Preserve no-op normal-mode behavior while still rejecting an in-progress cherry-pick.
- Require ACK-MANUAL-only normal runs to have a clean worktree and check out the target branch before reporting completion.
- Treat target-side `git revert <source-sha>` commits as active only when the revert itself is live on the target and changes the source commit paths.
- Treat no-footer and merge-only reapplications as superseding source-sha reverts when the target restores the source patch.
- Expand patch/liveness checks through target-side renames before deciding that side-merged or manually reapplied evidence is absent.
- Normalize target-side renamed patch paths back to source paths for patch-id comparison, while ignoring external diff drivers.
- Avoid treating side-branch commits that are merely reachable through `-s ours`/non-introducing merges as live target evidence.
- Keep stable version bump safeguards ahead of empty-commit and source-revert re-pick logic.
- Cache changed-path lookups during a planning run.
- Add focused regression coverage for the #4181 and #4261 review findings, including renamed target paths and no-footer reapplications.

## Addresses #4181 Review Threads

- `PRRT_kwDOAnNnU86MJ3kt`: no-op normal mode no longer checks out the target or rejects dirty worktrees when there is truly nothing to do, while still rejecting active cherry-picks.
- `PRRT_kwDOAnNnU86MJ3wG`: the quoted `-x` footer boundary is documented at the helper call site.
- `PRRT_kwDOAnNnU86MJ35H`: the spec load guard now explains why the script is loaded conditionally.
- `PRRT_kwDOAnNnU86MJ4fS`: reverted unrelated target merges no longer make older forward-ported fixes look absent.
- `PRRT_kwDOAnNnU86MJ4fb`: source-SHA reverts now force a re-pick only when the revert is live and not superseded.

## Addresses #4261 Review Threads

- `PRRT_kwDOAnNnU86MxWi4`: target-side renames are included when checking whether a side merge introduced a source commit's paths.
- `PRRT_kwDOAnNnU86MxWmF`: merge-only reapplications after source-sha reverts are recognized when the merge diff restores the source patch.
- `PRRT_kwDOAnNnU86MxX5_`: stable version bump guards run before empty-commit skipping, preserving the `MANUAL` gate.
- `PRRT_kwDOAnNnU86MxX6l`: the redundant pre-filter was removed; `target_commit_live_on_target?` remains the single liveness gate.
- `PRRT_kwDOAnNnU86MxX6_`: changed-path lookups are memoized per planning run.
- `PRRT_kwDOAnNnU86MxaW8`: ACK-MANUAL-only normal runs now verify a clean worktree and check out the target branch before completing.
- `PRRT_kwDOAnNnU86MxbRY`: source-SHA revert checks use rename-aware target paths, so a revert that removes a renamed target path is still treated as active.
- `PRRT_kwDOAnNnU86MxckF`: merge-only and no-footer patch comparisons normalize target renamed paths back to source paths and bypass external diff drivers.

## Validation

- `ruby -c script/release-forward-port && ruby -c react_on_rails/spec/react_on_rails/release_forward_port_script_spec.rb && git diff --check`
- `cd react_on_rails && bundle exec rspec spec/react_on_rails/release_forward_port_script_spec.rb` - 65 examples, 0 failures
- `cd react_on_rails && env GIT_TEST_FORCE_SIGNING_FALLBACK=1 bundle exec rspec spec/react_on_rails/release_forward_port_script_spec.rb` - 65 examples, 0 failures
- `BUNDLE_GEMFILE="$(git rev-parse --show-toplevel)/Gemfile" bundle exec rubocop script/release-forward-port react_on_rails/spec/react_on_rails/release_forward_port_script_spec.rb` - no offenses
- `script/ci-changes-detector origin/main` - routes RSpec tests + CI infrastructure/full optimized suite, no benchmarks
- pre-commit hook - trailing newline, RuboCop, autofix checks passed
- pre-push hook - branch Ruby lint and markdown-link routing passed
- `codex review --uncommitted` was run in bounded passes. It surfaced target-rename edge cases; those were fixed with regressions. The final retry timed out after exploratory reads, with no final finding list.

## Merge Rationale

This is a narrow post-merge hardening PR for the release forward-port helper after #4181. It reduces release-closeout risk by fixing known false-skip/false-pick cases, preserving manual handling for stable version bumps, and improving path/revert classification around side merges and target-side renames. The change is limited to `script/release-forward-port` and its spec, and every accepted review finding has targeted local regression coverage.

## Confidence Note

- Validated: syntax, whitespace, RuboCop, normal helper spec, forced-signing helper spec, CI detector, pre-commit hooks, pre-push hooks, optimized hosted CI, current-head review-thread triage, strict merge ledger.
- Evidence: branch head `68ef83ced`; hosted CI all green including `rspec-package-tests (4.0, latest, generators)` after 40m39s; CodeRabbit approved; `claude-review` passed; unresolved review threads count 0; `script/pr-merge-ledger 4261 --strict --changelog-classification not_user_visible --finding-dispositions <P2 not_applicable map> --pretty` returned `complete_allowed: true`.
- UNKNOWN: none.
- Residual risk: helper history classification remains conservative for unusual manually edited histories; late review cleanup-only/performance nits were auto-deferred to avoid restarting the final-candidate gate.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved `release-forward-port` handling for “no actionable work”: it now exits cleanly with **“Nothing to cherry-pick.”** and avoids unnecessary checks.
  * Tightened safety around dirty worktrees/cherry-pick-in-progress, and refined skip decisions for active revert/reapply scenarios (including more reliable treatment of `-x` evidence and rename cases).
  * Updated acknowledged manual flows to require a **clean worktree** and ensure the **target branch is checked out**.
* **Tests**
  * Expanded coverage for no-op runs, revert/reapply/rename edge cases, and `-x` evidence/supersession behavior across dry-run and revert-during-merge scenarios.

### CODE REVIEW

**Comment by justin808** [comment]
+ci-run-hosted

**Comment by justin808** [comment]
+ci-run-hosted

**Comment by justin808** [comment]
+ci-run-hosted

File: script/release-forward-port [hunk 1]
```diff
@@ -429,11 +511,47 @@ class ReleaseForwardPort
...
+
+    merge_changes_commit_paths?(first_parent, merge_sha, sha)
+  end
+
+  def merge_changes_commit_paths?(first_parent, merge_sha, sha)
+    paths = changed_paths_for_commit(sha)
+    return false if paths.empty?
+
+    diff_changes_paths?(first_parent, merge_sha, paths)
+  end
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:574]: Fixed in the current head. Merge path checks now use source_paths_for_target_diff / target_to_source_path_map so target-side renames are included before deciding side-merged evidence is absent. Regression coverage: skips side-merged -x reapplications when the target renamed the changed file before merge, plus the full helper spec is 65 examples passing.

File: script/release-forward-port [hunk 2]
```diff
@@ -362,6 +380,61 @@ class ReleaseForwardPort
...
+    end
+  end
+
+  def source_reapplied_without_x_after_revert?(sha, target, revert_sha)
+    source_patch_id = patch_id_for_commit(sha)
+
+    patch_equivalent_target_candidates(sha, target).any? do |target_sha|
+      patch_id_for_commit(target_sha) == source_patch_id &&
+        target_commit_introduced_after_revert?(target_sha, target, revert_sha) &&
+        target_commit_live_on_target?(target_sha, target)
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port]: Fixed. source_reapplied_by_merge_after_revert? now checks first-parent merge diffs and compares patch IDs after normalizing target paths back to source paths. Regression coverage includes merge-only reapplications after source-sha reverts, including the target-rename variant.

File: script/release-forward-port [hunk 3]
```diff
@@ -192,12 +200,15 @@ class ReleaseForwardPort
...
     return special_reason if special_reason
 
-    return "already forward-ported to #{target} via cherry-pick -x evidence with no later target revert" if
-      already_forward_ported_with_x?(sha, target)
-
     return "empty commit; cherry-picking would only create a no-op commit on #{target}" if
       patch_id_for_commit(sha) == EMPTY_PATCH_ID
 
+    return active_source_revert_skip_reason(subject, target, target_version) if
+      source_standard_revert_on_target?(sha, target)
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:216]: Fixed. Stable version-bump guards run before the empty-commit skip, preserving the MANUAL action. Regression coverage: keeps empty stable version bump commits manual.

File: script/release-forward-port [hunk 4]
```diff
@@ -362,6 +380,61 @@ class ReleaseForwardPort
     origin_sha && commit_on_target?(origin_sha, target) && !target_commit_live_on_target?(origin_sha, target)
   end
 
+  def source_standard_revert_on_target?(sha, target)
+    active_revert_shas =
+      standard_revert_shas_on_target(sha, target)
+      .reject { |revert_sha| standard_revert_on_target?(revert_sha, target) }
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port]: Fixed. The redundant reject was removed; target_commit_live_on_target? remains the single liveness gate for reverted target commits.

File: script/release-forward-port [hunk 5]
```diff
@@ -442,6 +560,10 @@ class ReleaseForwardPort
     git_lines("log", target, "--format=%H", "--no-merges", "--", *paths)
   end
 
+  def changed_paths_for_commit(sha)
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:645]: Fixed. changed_paths_for_commit now memoizes diff-tree results for the planning run.

File: script/release-forward-port [hunk 6]
```diff
@@ -85,6 +85,8 @@ class ReleaseForwardPort
       return 0
     end
 
+    return finish_no_action_plan(plan) unless actionable_plan?(plan)
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port]: Fixed in 456f4270c and retained in the current head. ACK-MANUAL-only normal runs now require a clean worktree and check out the target branch before reporting completion. Regression coverage: checks out the target branch before completing acknowledged manual-only plans.

File: script/release-forward-port [hunk 7]
```diff
@@ -362,6 +385,85 @@ class ReleaseForwardPort
...
+    active_revert_shas.any? do |revert_sha|
+      !source_revert_superseded_on_target?(sha, target, revert_sha)
+    end
+  end
+
+  def revert_changes_source_paths?(revert_sha, sha)
+    paths = changed_paths_for_commit(sha)
+    return false if paths.empty?
+
+    commit_changes_paths?(revert_sha, paths)
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port]: Fixed. revert_changes_source_paths? now expands source paths through target-side renames from the revert parent before checking whether the revert changed the source commit's paths. Regression coverage: picks source-sha reverts that remove a renamed target path.

File: script/release-forward-port [hunk 8]
```diff
@@ -362,6 +385,85 @@ class ReleaseForwardPort
...
+
+    patch_equivalent_target_candidates(sha, target).any? do |target_sha|
+      patch_id_for_commit(target_sha) == source_patch_id &&
+        target_commit_introduced_after_revert?(target_sha, target, revert_sha) &&
+        target_commit_live_on_target?(target_sha, target)
+    end
+  end
+
+  def source_reapplied_by_merge_after_revert?(sha, target, revert_sha)
+    source_patch_id = patch_id_for_commit(sha)
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:440]: Fixed. Patch-id comparisons for merge-only and no-footer target reapplications now normalize target renamed paths back to source paths, and patch inputs use --no-ext-diff. Regression coverage includes merge-only reapplications after target renames, no-footer reapplications after target renames, and direct target patches that rename the source path.

File: script/release-forward-port [hunk 9]
```diff
@@ -362,6 +391,85 @@ class ReleaseForwardPort
...
+      .select { |revert_sha| revert_changes_source_paths?(revert_sha, sha) }
+    return false if active_revert_shas.empty?
+
+    active_revert_shas.any? do |revert_sha|
+      !source_revert_superseded_on_target?(sha, target, revert_sha)
+    end
+  end
+
+  def revert_changes_source_paths?(revert_sha, sha)
+    paths = changed_paths_for_commit(sha)
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port]: Fixed. Source-sha revert detection now includes target-side renamed paths via source_paths_for_target_diff, and the renamed-revert regression is covered by the helper spec.

File: script/release-forward-port [hunk 10]
```diff
@@ -442,17 +595,48 @@ class ReleaseForwardPort
...
   end
 
+  def source_paths_for_target_diff(first_parent, sha)
+    paths = changed_paths_for_commit(sha)
+    return [] if paths.empty?
+
+    (paths + renamed_target_paths_for_source_paths(first_parent, sha, paths)).uniq
+  end
+
+  def renamed_target_paths_for_source_paths(first_parent, sha, paths)
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port]: Fixed. The renamed target path lookup is now memoized by [first_parent, source_sha] during the planning run.

File: script/release-forward-port [hunk 11]
```diff
@@ -192,21 +206,38 @@ class ReleaseForwardPort
...
+  end
+
+  def patch_equivalent_skip_reason(sha, target, patch_equivalent_commits)
+    return unless patch_equivalent_commits[sha] &&
+                  patch_equivalent_commit_present_on_target?(sha, target)
+
+    "patch already exists on #{target} according to target history with no later standard revert"
+  end
 
+  def active_source_revert_skip_reason(subject, target, target_version)
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:235]: [auto-deferred]
This is a clarity-only cleanup: the current nil-return behavior is intentional and covered by the source-sha revert tests. Deferring to avoid restarting the current-head CI/review gate for a non-correctness refactor.

File: script/release-forward-port [hunk 12]
```diff
@@ -339,11 +370,9 @@ class ReleaseForwardPort
 
   def already_forward_ported_with_x?(sha, target)
     return true if source_cherry_pick_origin_on_target?(sha, target)
+    return false if source_standard_revert_on_target?(sha, target)
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:373]: [auto-deferred]
This is a performance-only duplicate guard cleanup. Current behavior is correct and covered; deferring to avoid restarting the current-head CI/review gate for optional churn.

File: script/release-forward-port [hunk 13]
```diff
@@ -429,11 +546,47 @@ class ReleaseForwardPort
   end
 
   def target_merge_descendants(sha, target)
-    git_lines("log", target, "--merges", "--ancestry-path", "#{sha}..#{target}", "--format=%H")
+    git_lines("log", target, "--merges", "--ancestry-path", "#{sha}..#{target}", "--format=%H").select do |merge_sha|
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:553]: [auto-deferred]
This is a performance-only lazy-enumerator optimization. Current behavior is correct and locally validated; deferring to avoid restarting the current-head CI/review gate.

File: script/release-forward-port [hunk 14]
```diff
@@ -442,17 +595,48 @@ class ReleaseForwardPort
...
+  def renamed_target_paths_for_source_paths(first_parent, sha, paths)
+    merge_base = git("merge-base", first_parent, sha).strip
+    source_paths = paths.to_h { |path| [path, true] }
+
+    git_lines("diff", "--name-status", "--find-renames", merge_base, first_parent).filter_map do |line|
+      status, old_path, new_path = line.split("\t", 3)
+      new_path if status.start_with?("R") && new_path && source_paths.key?(old_path)
+    end.uniq
+  end
+
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:644]: [auto-deferred]
This is a performance-only patch-id cache optimization. Current behavior is correct and covered by targeted regressions; deferring to avoid optional churn on the final candidate.

File: script/release-forward-port [hunk 15]
```diff
@@ -413,13 +521,22 @@ class ReleaseForwardPort
...
   def target_commit_reverted_through_merge?(sha, target)
-    # Intentionally conservative: if any merge that included this commit was
-    # reverted, treat the commit as not live. Re-picking a present change is
-    # safer than skipping a release fix that the target tree no longer has.
+    # Only merge reverts that introduced this commit can make it non-live. Later
+    # unrelated merges often descend from already-forward-ported fixes through
+    # first-parent history; their reverts should not make an older fix look absent.
     @target_merge_revert_cache ||= {}
     cache_key = [sha, target]
     return @target_merge_revert_cache.fetch(cache_key) if @target_merge_revert_cache.key?(cache_key)
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:546]: [auto-deferred]
This is a performance-only target-merge-descendant cache optimization. Current behavior is correct and locally validated; deferring to avoid restarting current-head gates.

File: script/release-forward-port [hunk 16]
```diff
@@ -339,11 +370,9 @@ class ReleaseForwardPort
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:370]: [auto-deferred]
This duplicates the source_standard_revert_on_target? memoization/performance concern. Current behavior is correct; deferring optional optimization to avoid current-head churn.

File: script/release-forward-port [hunk 17]
```diff
@@ -192,21 +206,38 @@ class ReleaseForwardPort
...
+  end
 
+  def active_source_revert_skip_reason(subject, target, target_version)
+    # A target-side `git revert <source-sha>` names the release commit rather
+    # than the earlier target cherry-pick, so do not trust either -x footer or
+    # patch-equivalence evidence while that source revert is still active.
+    # Version-bump guards still win because reapplying release-only version files
+    # to the target branch is never an automatic forward-port.
     version_bump_skip_reason(subject, target, target_version)
   end
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:242]: [auto-deferred]
This is a clarity-only cleanup around an intentional nil return. Current behavior is correct and covered; deferring to avoid restarting the current CI/review gate for non-correctness cleanup.

File: script/release-forward-port (hunk 12 above)
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:373]: [auto-deferred]
This duplicates the source_standard_revert_on_target? memoization/performance concern. Current behavior is correct; deferring optional optimization to avoid current-head churn.

File: script/release-forward-port [hunk 18]
```diff
@@ -429,11 +546,47 @@ class ReleaseForwardPort
...
+
+    first_parent, *merged_side_parents = parents
+    return false if commit_descends_from?(first_parent, sha)
+    return false unless merged_side_parents.any? { |parent| commit_descends_from?(parent, sha) }
+
+    merge_changes_commit_paths?(first_parent, merge_sha, sha)
+  end
+
+  def merge_changes_commit_paths?(first_parent, merge_sha, sha)
+    paths = source_paths_for_target_diff(first_parent, sha)
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:570]: [auto-deferred]
This is a performance-only fast-path suggestion. Current rename-aware behavior is correct and covered by regression tests; deferring to avoid optional churn on the final candidate.

File: script/release-forward-port [hunk 19]
```diff
@@ -362,6 +391,88 @@ class ReleaseForwardPort
...
+        patch_id_for_target_candidate(target_sha, sha) == source_patch_id
+    end
+  end
+
+  def source_reapplied_by_merge_after_revert?(sha, target, revert_sha)
+    source_patch_id = patch_id_for_commit(sha)
+
+    target_first_parent_merges_after(revert_sha, target).any? do |merge_sha|
+      target_commit_live_on_target?(merge_sha, target) &&
+        patch_id_for_merge_source_paths(merge_sha, sha) == source_patch_id
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:444]: Declining this one intentionally. Detecting a source patch as a subset of a same-file merge conflict-resolution diff would require partial patch matching, and extra same-path hunks are ambiguous release-closeout evidence. The current behavior is conservative: it only treats the merge as superseding the source-sha revert when the source patch comparison is exact after path normalization; otherwise it forces a re-pick/manual conflict instead of silently skipping a release commit.

File: script/release-forward-port [hunk 20]
```diff
@@ -429,30 +550,215 @@ class ReleaseForwardPort
...
+    @renamed_target_path_pairs_cache[cache_key] =
+      rename_base_refs.flat_map do |from_ref|
+        renamed_target_path_pairs_from_ref(from_ref, first_parent, source_paths)
+      end.uniq
+  end
+
+  def renamed_target_path_pairs_from_ref(from_ref, first_parent, source_paths)
+    git_lines("diff", "--name-status", "--find-renames", from_ref, first_parent).filter_map do |line|
+      status, old_path, new_path = line.split("\t", 3)
+      [old_path, new_path] if status.start_with?("R") && new_path && source_paths.key?(old_path)
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:641]: Fixed in 68ef83ced. renamed_target_path_pairs_from_ref now uses --find-renames=20%, matching the lower threshold used by normalized patch-id comparisons. Validation: focused rename cases, syntax/diff check, RuboCop, and both full helper spec runs passed (65 examples each).

File: script/release-forward-port (hunk 12 above)
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:373]: [auto-deferred]
Performance-only duplicate history lookup noted. Deferring to avoid restarting the final-candidate gate; current behavior is already covered by the revert/reapply regression tests.

File: script/release-forward-port [hunk 21]
```diff
@@ -408,18 +519,28 @@ class ReleaseForwardPort
...
+        target_commit_live_on_target?(target_sha, target)
     end
   end
 
   def target_commit_live_on_target?(sha, target)
-    !standard_revert_on_target?(sha, target) && !target_commit_reverted_through_merge?(sha, target)
+    return false if standard_revert_on_target?(sha, target)
+    return false if target_commit_reverted_through_merge?(sha, target)
+
+    commit_on_first_parent?(sha, target) || target_merge_descendants(sha, target).any?
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:531]: [auto-deferred]
Performance-only duplicate merge-descendant traversal noted. Deferring to avoid restarting the final-candidate gate; current behavior is already covered and hosted CI is green.

File: script/release-forward-port [hunk 22]
```diff
@@ -429,30 +550,215 @@ class ReleaseForwardPort
...
+    @renamed_target_path_pairs_cache[cache_key] =
+      rename_base_refs.flat_map do |from_ref|
+        renamed_target_path_pairs_from_ref(from_ref, first_parent, source_paths)
+      end.uniq
+  end
+
+  def renamed_target_path_pairs_from_ref(from_ref, first_parent, source_paths)
+    git_lines("diff", "--name-status", "--find-renames=20%", from_ref, first_parent).filter_map do |line|
+      status, old_path, new_path = line.split("\t", 3)
+      [old_path, new_path] if status.start_with?("R") && new_path && source_paths.key?(old_path)
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:641]: Not applicable on current head: `renamed_target_path_pairs_from_ref` already uses `git diff --name-status --find-renames=20%`, matching `rename_detection_option` for normalized patch diffs. No code change needed.

File: script/release-forward-port [hunk 23]
```diff
@@ -362,6 +391,88 @@ class ReleaseForwardPort
...
+  end
+
+  def forwarded_shas_with_x(sha, target)
+    footer = format(CHERRY_PICK_FOOTER_FORMAT, sha)
+    # This history search can false-positive if a later target commit quotes the
+    # footer verbatim. The live-target check by callers mitigates reverted target
+    # copies; the runbook documents the remaining quoted-footer boundary.
+    git_lines("log", target, "--fixed-strings", "--grep", footer, "--format=%H")
+  end
+
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:475]: [auto-deferred]
Performance-only memoization cleanup noted. Deferring to avoid restarting the final-candidate gate; current behavior is already covered and hosted CI is green.

File: script/release-forward-port [hunk 24]
```diff
@@ -429,30 +550,215 @@ class ReleaseForwardPort
...
+    normalize_unprefixed_patch_path_line(line, target_to_source_paths)
+  end
+
+  def normalize_patch_diff_header(line, target_to_source_paths)
+    trailing_newline = line.end_with?("\n") ? "\n" : ""
+    header = line.delete_prefix("diff --git ").delete_suffix("\n")
+
+    sorted_paths = target_to_source_paths.sort_by { |target_path, _source_path| -target_path.length }
+    sorted_paths.each do |target_path, source_path|
+      next if target_path == source_path
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:709]: [auto-deferred]
Clarity-only unreachable-guard cleanup noted. Deferring to avoid restarting the final-candidate gate; it does not change behavior.

File: script/release-forward-port [hunk 25]
```diff
@@ -429,30 +550,215 @@ class ReleaseForwardPort
...
+    return normalize_prefixed_patch_path_line(line, target_to_source_paths) if line.start_with?("--- ", "+++ ")
+
+    normalize_unprefixed_patch_path_line(line, target_to_source_paths)
+  end
+
+  def normalize_patch_diff_header(line, target_to_source_paths)
+    trailing_newline = line.end_with?("\n") ? "\n" : ""
+    header = line.delete_prefix("diff --git ").delete_suffix("\n")
+
+    sorted_paths = target_to_source_paths.sort_by { |target_path, _source_path| -target_path.length }
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port:707]: [auto-deferred]
Micro-optimization/clarity cleanup noted. Deferring to avoid restarting the final-candidate gate; exact header matching makes this non-behavioral.
